arlo_server/jurisdictions.py

```diff
@@ -1,35 +1,139 @@
+        .group_by(Batch.jurisdiction_id)
+        .values(Batch.jurisdiction_id, func.count())
+    )
+    audited_ballot_count_by_jurisdiction = dict(
+        SampledBallotDraw.query.filter_by(round_id=round.id)
+        .join(SampledBallot)
+        .filter(SampledBallot.vote.isnot(None))
+        .join(Batch)
+        .group_by(Batch.jurisdiction_id)
+        .values(Batch.jurisdiction_id, func.count())
+    )
+
+    def num_ballots_sampled(jurisdiction_id: str) -> int:
+        return sampled_ballot_count_by_jurisdiction.get(jurisdiction_id, 0)
+
+    # NOT_STARTED = the jurisdiction hasn’t set up any audit boards
+    # IN_PROGRESS = the audit boards are set up
+    # COMPLETE = all of the audit boards have signed off on their ballots
+    def status(jurisdiction_id: str) -> JurisdictionStatus:
+        num_set_up = audit_boards_set_up.get(jurisdiction_id, 0)
+        num_signed_off = audit_boards_signed_off.get(jurisdiction_id, 0)
+        num_sampled = num_ballots_sampled(jurisdiction_id)
+
+        # Special case: jurisdictions that don't get any ballots assigned are
+        # COMPLETE from the get-go
+        if num_sampled == 0:
+            return JurisdictionStatus.COMPLETE
+        elif num_set_up == 0:
+            return JurisdictionStatus.NOT_STARTED
+        elif num_signed_off != num_set_up:
+            return JurisdictionStatus.IN_PROGRESS
+        else:
+            return JurisdictionStatus.COMPLETE
+
+    def num_ballots_audited(jurisdiction_id: str) -> int:
+        if online:
+            return audited_ballot_count_by_jurisdiction.get(jurisdiction_id, 0)
+        else:
+            # For offline audits, we can't track incremental progress on
+            # auditing ballots, so we just report 0 ballots until the round is
+            # complete, then report all the sampled ballots.
+            return (
+                0
+                if status(jurisdiction_id) != JurisdictionStatus.COMPLETE
+                else num_ballots_sampled(jurisdiction_id)
+            )
+
+    return {
+        j.id: {
+            "status": status(j.id),
+            "numBallotsSampled": num_ballots_sampled(j.id),
+            "numBallotsAudited": num_ballots_audited(j.id),
+        }
+        for j in jurisdictions
     }
 
 
 @app.route("/election/<election_id>/jurisdiction", methods=["GET"])
 @with_election_access(UserType.AUDIT_ADMIN)
 def list_jurisdictions(election: Election):
-    jurisdictions = (
-        Jurisdiction.query.filter_by(election_id=election.id)
-        .order_by(Jurisdiction.name)
-        .all()
+    jurisdictions = sorted(election.jurisdictions, key=lambda j: j.name)
+
+    current_round = get_current_round(election)
+    round_status = round_status_by_jurisdiction(
+        current_round, jurisdictions, election.online
     )
 
-    return jsonify([serialize_jurisdiction(j) for j in jurisdictions])
+    json_jurisdictions = [
+        serialize_jurisdiction(j, round_status[j.id]) for j in jurisdictions
```

Comment:
is there ever a case where we're not generating the `round_status`, and should that generation be within the `serialize_jurisdiction` so as not to have a potential mismatch between the jurisdiction and corresponding status?

Reply:
`round_status_by_jurisdiction()` generates the round statuses for all the jurisdictions in bulk, while `serialize_jurisdiction` only looks at one jurisdiction at a time. The only way I can think of to make the round status generation within `serialize_jurisdiction` is to memoize `round_status_by_jurisdiction()`, but that seems more confusing/error-prone in this case. Do you have other ideas?